File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-4.java
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) return null;

        PriorityQueue<ListNode> minHeap = new PriorityQueue<>((a, b) -> a.val - b.val);
        for (var list : lists) {
            if (list != null) {
                minHeap.offer(list);
            }
        }

        ListNode dummy = new ListNode(0);
        ListNode ptr = dummy;

        while (!minHeap.isEmpty()) {
            ListNode node = minHeap.poll();

            ptr.next = node;
            ptr = ptr.next;

            if (node.next != null) {
                minHeap.offer(node.next);
            }
        }

        return dummy.next;
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-4.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.size() == 0) {
            return nullptr;
        }

        auto cmp = [](ListNode* a, ListNode* b) {
            return a->val > b->val;
        };

        priority_queue<ListNode*, vector<ListNode*>, decltype(cmp)> pq(cmp);

        for (auto list : lists) {
            if (list) {
                pq.push(list);
            }
        }

        ListNode* dummy = new ListNode();
        ListNode* currNode = dummy;

        while (!pq.empty()) {
            ListNode* minNode = pq.top();
            currNode->next = minNode;
            pq.pop();

            if (minNode->next) {
                pq.push(minNode->next);
            }

            currNode = currNode->next;
        }

        return dummy->next;
    }
};
